magento/magento2
# Fix PageCache: async rendering of blocks can corrupt layout cache #8554 #9050

**Description**
This pull request expects to fix layout cache corruption due to async pagecache requests, when the async request generates a layout cacheId based on same handles as a cms page, but without loading the cms page associated, so cms page related layout updates get lost. Further explanation on issues #8554 and #9050. The testing scenario may sound weird, but it actually happens due to layout cache expiration, so layout cache expires but full page cache is still valid.

**Fixed Issues**
1. magento/magento2#8554: PageCache: async rendering of blocks can corrupt layout cache
2. magento/magento2#9050: Randomly getting an empty body that gets cached

**Manual testing scenarios**
1. Use the cms home page to add a block via Layout Update XML. (I've also experienced issued with page layout setting not being used).
2. Ensure layout and full page cache are enabled, at least this two.
3. Visit cms page on frontend, so layout cache and full page cache are generated.
4. Clean layout cache: bin/magento cache:clean layout
5. curl -gI "http://local.mg2develop.com/page_cache/block/render/?ajax=1&blocks=[%22header.login%22]&handles=[%22default%22%2C%22cms_index_index%22%2C%22cms_page_view%22%2C%22cms_index_index_id_home%22]&originalRequest=a&version="
6. Clean only full page cache: bin/magento cache:clean full_page

**Adapted Tests**
\Magento\PageCache\Test\Unit\Controller\Block\RenderTest::testExecute
\Magento\PageCache\Test\Unit\Controller\Block\EsiTest::testExecute

**Expected Result**
Page rendered correctly, as the first time.

**Actual Result**
Depending on included content, full blank page may be returned, but at least, it fails to apply page layout update handles, so block added via layout update is not shown.
